regression/contracts/variant_multi_instruction_loop_head/test.desc

```diff
@@ -0,0 +1,19 @@
+CORE
+main.c
+--apply-loop-contracts
+^VERIFICATION FAILED$
+^EXIT=10$
+^SIGNAL=0$
+--
+--
+This test fails even without the fix proposed in the commit, so it should be improved.
```

Comment:
What is the observable difference in this test of the fix?  If none then ... can we have a test case that does have an observable difference when fixed?

Presumably it will need a short-cut operator with one operation that alters a value?  What about a fancy `for-else` loop like this:

```
int i = 0;
int i_valid_element = 1;

while (i < limit || i_valid_element = 0) {
  if (a[i] == target) {
    break;
  }
  ++i;
}

if (!i_valid_element) {
  realloc(a, ... );
  a[i] = target;
}

return i;
```

Reply:
> What is the observable difference in this test of the fix?

The difference is that the old value of the loop variant was recorded in the middle of loop head instructions. Simply doing `std::next(loop_head)` isn't sufficient to get to the body, where it should be recorded.

The `--show-goto-functions` on the test would show the difference, but there is no observation failure in the CBMC output. I could try to come up with a test where it would lead to a failure, though.

Reply:
If you have the time / energy it would be nice.  Adding test cases that don't have an observable difference is always a little iffy.  I appreciate that the test case you add would cover the source lines though.

It is pretty synthetic but...

```
uint64_t cookie = 0;
int connection_working = 0;
while (!connection_working && (cookie != 0 || cookie = get_new_cookie())) {
  connection_working = connect(cookie);
}
```

`get_new_cookie()` has a post condition that `cookie` is valid (i.e. non-zero) and `connect()` has a pre condition that `cookie` is valid as well.  It is reasonable for the user to think that cookie is guaranteed to be valid and invariant in the loop body.